File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
    private:
    vector<string>vec;
    bool check(string&str)
    {
        stack<char>st;
        for(auto&x:str)
        {
            if(st.empty()&&x==')')return false;
            if(x=='(')st.push(x);
            else st.pop();
        }
        if(st.empty())return true;
        return false;
    }
    void func(int i,int n,string &str)
    {
        if(i==n*2)
        {
            if(check(str))
            vec.push_back(str);
            return ;
        }
        if(i>n*2)return ;
        str+='(';
        func(i+1,n,str);
        str.pop_back();
        str+=')';
        func(i+1,n,str);
         str.pop_back();
    }
public:
    vector<string> generateParenthesis(int n) {
        string str="";
        func(0,n,str);
        return vec;
    }
};

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public void func(int open,int close, int n,String str,List<String>res)
    {
        if(open == n && close == n ){
        res.add(str);
        return;
        }
        else if(open > n) return ;
        else if(close>open)return ;
        func(open+1,close,n,str+"(",res);
        func(open,close+1,n,str+")",res);

    }
    public List<String> generateParenthesis(int n) {

        List<String> res=new ArrayList<>();
        func(0,0,n,"",res);
        return res;
    }
}